Test and further document ToBaseId

MaxGabriel opened the pull request:
My coworker Lev was adding this typeclass to our codebase and we hadn't used it before. I added a little more documentation that I think would help clarify things, particularly what the point of the witness function was. More importantly I added a test for this typeclass.

Before submitting your PR, check that you've:

- [ ] Bumped the version number -- Note: I saw there were other unreleased changes, so I just added an "unreleased" section to the changelog with my changes. I can bump the version number if you'd like
- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html) (No new APIs)
- [ ] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock (No new APIs)

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [ ] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

<!---Thanks so much for contributing! :)

_If these checkboxes don't apply to your PR, you can delete them_

If you're unsure on what the new version number should be, feel free to ask.

-->

MaxGabriel commented:
![image](https://user-images.githubusercontent.com/1274145/85208884-cd5d3b80-b301-11ea-98e6-69c14674bc4b.png)

MaxGabriel commented on changelog.md:
```diff
@@ -1,3 +1,8 @@
+Unreleased
+========
+- @maxgabriel
+  - [#190](https://github.com/bitemyapp/esqueleto/pull/190) Further dcument and test `ToBaseId`
```
```suggestion
  - [#190](https://github.com/bitemyapp/esqueleto/pull/190) Further document and test `ToBaseId`
```

MaxGabriel commented on src/Database/Esqueleto/Internal/Internal.hs:
```diff
@@ -1007,12 +1007,13 @@ case_ = unsafeSqlCase
 --
--- For this example, declare:
+-- In this example, Bar is said to be the BaseEnt(ity), and Foo the child.
+-- To model this in Esqueleto, declare:
```
Explicitly clarifying what is supposed to be the Base was helpful to me. It's kind of confusing b/c the code reads as `BaseEnt Foo`, which vaguely looks like Foo is the BaseEnt

MaxGabriel commented on src/Database/Esqueleto/Internal/Internal.hs:
```diff
@@ -1007,12 +1007,13 @@ case_ = unsafeSqlCase
 -- instance ToBaseId Foo where
 --   type BaseEnt Foo = Bar
---   toBaseIdWitness = FooKey
+--   toBaseIdWitness barId = FooKey barId
```
Lev and I thought removing the eta reduction makes it more clear what this function is doing

MaxGabriel commented on src/Database/Esqueleto/Internal/Internal.hs:
```diff
@@ -1371,7 +1372,11 @@ instance SqlString a => SqlString (Maybe a) where
   type BaseEnt ent :: *
+  -- | Convert from the key of the BaseEnt(ity) to the key of the child entity.
+  -- This function is not actually called, but that it typechecks proves this operation is safe.
+  -- You may be able to succinctly implement this with 'Data.Coerce.coerce'
```
I think coerce is both a nice implementation tip, but also gives good intuition for what this function is about

MaxGabriel replied:
Hm, after thinking about it, it's probably a mistake to recommend `coerce`. Since:

```
ghci> :t coerce (UserKey UUID.nil) :: OrganizationId
coerce (UserKey UUID.nil) :: OrganizationId :: OrganizationId
```

is just fine, but is actually incorrect

parsonsmatt replied:
Yeah, `coerce` is a little unsafe, in that it *does* work but it relies on the user guaranteeing that the two underlying IDs are indeed the same.

Leaving it in there is probably fine but we should leave a warning to the edge case.

MaxGabriel commented:
Ok cool, I removed coerce b/c it seems like bad advice to give people. I also rebased with master which resolved a conflict.

I saw that 3.3.1 was released from the last PR—would you like this to be 3.3.1.1, or just mark this as "unreleased" in the changelog

parsonsmatt commented:
I'm happy releasing this as 3.3.3.2

MaxGabriel commented:
Ok bumped to 3.3.3.2
